### ISSUE TITLE
How to predict the topic of a new document?

### ISSUE THREAD

**Issue by gongshaojie12** [MaartenGr/BERTopic#310]
When I use the `transform` method to make topic predictions for new documents after the topic model training is completed, but the predicted topics are the topics included in the model training data, which is obviously wrong. How can I use the trained model for topic extraction of new documents? Thank you very much!

**Comment by MaartenGr** [comment]
This is actually what you should expect from any `transform` method. When you fit any model on data, some representation of the data gets saved in the model. Whenever you then predict or transform new data, you are predicting how that new data relates to the data you originally trained on. 

With BERTopic, it is the same. You `fit` on data and learn the topics from those. Whenever you then `transform` new data, they will be modeled to the topics learned from the original data. Thus, this behavior is not wrong but exactly what a `transform` method should do. 

In other words, whenever you apply `transform` it is indeed extracting topics for those new documents but previously unseen topics will not be generated. If you want to generate new topics, then you will have to `fit` on all data, including the new data.

**Comment by gongshaojie12** [comment]
So how to generate unseen topics in real time? It is impossible to train with historical data when encountering an unseen topic, so that it cannot meet the requirements of real-time.

**Comment by MaartenGr** [comment]
BERTopic can still work on relatively real-time data as long as you expect the topics to remain relatively stable over time. When you notice they are changing, simply re-train the model in between and swap models the moment the newer model has been trained. 

Having said that, if you are looking for a model that outputs different keywords/topics without having been trained on any data, I would advise looking at [KeyBERT](https://github.com/MaartenGr/KeyBERT) or [YAKE](https://github.com/LIAAD/yake) instead. These keyword extraction models can generate related words on the fly. So I believe they would fit your requirements.

**Comment by gongshaojie12** [comment]
Thank you very much for your reply.

**Comment by MaartenGr** [comment]
No problem, glad I could be of help!
